eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -408,55 +409,74 @@ object Helpers {
-      val mainDelayedTx = generateTx("main-delayed-output")(Try {
-        val claimDelayed = Transactions.makeClaimDelayedOutputTx(tx, Satoshi(localParams.dustLimitSatoshis), localRevocationPubkey, remoteParams.toSelfDelay, localDelayedPubkey, localParams.defaultFinalScriptPubKey, feeratePerKwDelayed)
-        val sig = keyManager.sign(claimDelayed, keyManager.delayedPaymentPoint(localParams.channelKeyPath), localPerCommitmentPoint)
-        Transactions.addSigs(claimDelayed, sig)
-      })
+      val mainDelayedTx = makeDelayedTx(tx, "main-delayed-output")
 
       // those are the preimages to existing received htlcs
-      val preimages = commitments.localChanges.all.collect { case u: UpdateFulfillHtlc => u.paymentPreimage }
-
-      val htlcTxes = localCommit.publishableTxs.htlcTxsAndSigs.collect {
-        // incoming htlc for which we have the preimage: we spend it directly
-        case HtlcTxAndSigs(txinfo@HtlcSuccessTx(_, _, paymentHash), localSig, remoteSig) if preimages.exists(r => sha256(r) == paymentHash) =>
-          generateTx("htlc-success")(Try {
-            val preimage = preimages.find(r => sha256(r) == paymentHash).get
-            Transactions.addSigs(txinfo, localSig, remoteSig, preimage)
-          })
-
-        // (incoming htlc for which we don't have the preimage: nothing to do, it will timeout eventually and they will get their funds back)
-
-        // outgoing htlc: they may or may not have the preimage, the only thing to do is try to get back our funds after timeout
-        case HtlcTxAndSigs(txinfo: HtlcTimeoutTx, localSig, remoteSig) =>
-          generateTx("htlc-timeout")(Try {
-            Transactions.addSigs(txinfo, localSig, remoteSig)
-          })
-      }.flatten
-
-      // all htlc output to us are delayed, so we need to claim them as soon as the delay is over
-      val htlcDelayedTxes = htlcTxes.flatMap {
-        txinfo: TransactionWithInputInfo =>
-          generateTx("claim-htlc-delayed")(Try {
-            val claimDelayed = Transactions.makeClaimDelayedOutputTx(
-              txinfo.tx,
-              Satoshi(localParams.dustLimitSatoshis),
-              localRevocationPubkey,
-              remoteParams.toSelfDelay,
-              localDelayedPubkey,
-              localParams.defaultFinalScriptPubKey, feeratePerKwDelayed)
-            val sig = keyManager.sign(claimDelayed, keyManager.delayedPaymentPoint(localParams.channelKeyPath), localPerCommitmentPoint)
-            Transactions.addSigs(claimDelayed, sig)
-          })
+      val preimages = commitments.localChanges.all.collect { case u: UpdateFulfillHtlc => sha256(u.paymentPreimage) -> u.paymentPreimage }.toMap
+      val inFlightHtlcs = commitments.localCommit.spec.htlcs.map(htlc => htlc.add.paymentHash -> htlc.add.amountMsat).toMap
+      var onChainHashes = Set.empty[BinaryData] // this is needed to collect hashes of payments which can be resolved on-chain
+
+      val successAndDelayedTxs = for {
```

Comment:
An advantage of using `for`-sequences here:

- if either `success` or `delayed` fails then no tx is published as opposed to current approach where `success` may be published but `delayed` won't be (because applying on-chain fees would make it dusty for example).
- if either `success` or `delayed` fails then no `PaymentSettlingOnChain` will be sent, but `PaymentLostOnChain` will be sent for this payment instead.

Further, if `LocalCommitPublished` can be modified to store `(success, delayed)` and `(timeout, delayed)` tuples directly then UI and API may inform how much blocks is left for each delayed payment: `CLTV(timeout) + CSV(delayed)` as well as total on-chain fees paid: `timeout.input.amount - delayed.output.amount`.